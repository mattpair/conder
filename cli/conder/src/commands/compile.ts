
import { GluegunToolbox, GluegunCommand } from 'gluegun'
import {string_to_environment} from 'conduit_compiler'  

const command: GluegunCommand = {
  name: 'compile',
  alias: ['c'],
  run: async (toolbox: GluegunToolbox) => {
    const {
      parameters,
      print: { info, error },
      filesystem,
    } = toolbox
    
    const deployment_name = parameters.first === undefined ? "app" : parameters.first
    const artifact_name = `${deployment_name}.json`
    info(`Writing artifacts to ${artifact_name}`)
    
    const conduit = filesystem.read("main.cdt")
    if (conduit === undefined) {
      error(`Could not find required file main.cdt`)
      process.exit(1)
    }
    const output = string_to_environment(conduit)
    switch (output.kind) {
      case "success":
        filesystem.write(artifact_name, JSON.stringify(output.env))
        break
      case "error":
        error(output.reason)
        process.exit(1)
    }
  }
}

module.exports = command
