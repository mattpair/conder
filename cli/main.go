package main

import (
	// "./cmd"
	"fmt"
	"io/ioutil"
	"net/http"
)

func main() {
	out, err := http.Get("https://localhost:7344")
	if err != nil {
		fmt.Println("error", err)
		return
	}
	defer out.Body.Close()
	if out != nil {
		fmt.Println(out)
	}
	body, err := ioutil.ReadAll(out.Body)

	if err != nil {
		fmt.Println("error", err)
	}
	if body != nil {
		fmt.Println(out)
	}

	// cmd.Execute()
}
