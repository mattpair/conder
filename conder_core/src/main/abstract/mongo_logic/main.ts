import { ActionSequence, calculate_lock_requirements, LockRequirements } from './lock_calculation';
import { MongoNodeSet } from '../globals/mongo';
import { TargetNodeSet, NodeSet } from "../IR";
import { ScopeMap } from '../../data_structures/scope_map';
import { MONGO_ACTION_SUMMARIZER } from './action_summarizer';

type LockCalculator = (input: Map<string, TargetNodeSet<MongoNodeSet>[]>) =>  Map<string, LockRequirements[string]>

export const MONGO_LOCK_CALCULATOR: LockCalculator = input => {
    
    const actions: Record<string, ActionSequence> = {}
    input.forEach((v, k) => {
        actions[k] = MONGO_ACTION_SUMMARIZER(v)
    })
    const locks = calculate_lock_requirements(actions)
    
    return new Map(Object.entries(locks))
}

