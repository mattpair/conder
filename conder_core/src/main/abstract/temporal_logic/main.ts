type Actions = {
    get: {id: string},
    mutation: {id: string, using: string[]},
}
type ActionKind = keyof Actions
type AnyAction = {
    [K in keyof Actions]: {kind: K} & Actions[K]
}[keyof Actions]


export type ActionSequence = AnyAction[]
export type LockRequirements = Record<string, {global: string, kind: "r" | "w"}[]>

export function calculate_lock_requirements(sequences: Record<string, ActionSequence>): LockRequirements {
    const actionAgainstData = new Map<string, Set<ActionKind>>()
    const lockReqs: LockRequirements = {}

    Object.values(sequences).forEach(seq => {
        seq.forEach(action => {
            const ops = actionAgainstData.get(action.id)
            if (ops === undefined) {
                actionAgainstData.set(action.id, new Set<ActionKind>([action.kind]))
            } else {
                ops.add(action.kind)
                actionAgainstData.set(action.id, ops)
            }
        })
    })
    
    Object.keys(sequences).forEach(func => {
        lockReqs[func] = []
        const previouslyGot = new Set<string>()
        sequences[func].forEach(action => {
            
            switch (action.kind) {
                case "get":
                    if (previouslyGot.has(action.id)) {
                        if (actionAgainstData.get(action.id).has("mutation")) {
                            lockReqs[func].push({global: action.id, kind: "r"})
                        }
                    } else {
                        previouslyGot.add(action.id)
                    }
                case "mutation":
                    break
                default: 
                    const n: never = action
            }
        })
    })

    return lockReqs
}