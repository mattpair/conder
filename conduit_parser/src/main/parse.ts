import { Symbol } from './lexicon';
import { assertNever } from './utils';
import { FileLocation } from "./utils";
import * as common from './entity/basic'


export namespace Parse {
    
    export type File = 
    common.Entity<"File"> & 
    common.ParentOfMany<Struct> &  
    common.ParentOfMany<common.Enum> & 
    common.ParentOfMany<Function> &
    common.ParentOfMany<StoreDefinition> &
    {readonly loc: FileLocation}

    export type CustomTypeEntity = common.IntrafileEntity<"CustomType", {type: string, modification: common.TypeModification}>
    export type FieldType = common.BaseFieldType<() => CustomTypeEntity>
    export type Field = common.BaseField<FieldType>

    export type VariableReference = common.IntrafileEntity<"VariableReference", {val: string}>
    export type Append = common.IntrafileEntity<"Append", {storeName: string, variableName: string}>
    export type Nothing = common.IntrafileEntity<"Nothing", {}>
    export type Returnable = common.PolymorphicEntity<"Returnable", () => Nothing | Assignable>
    export type ReturnStatement = common.IntrafileEntity<"ReturnStatement", common.RequiresOne<Returnable>>
    export type Assignable = common.PolymorphicEntity<"Assignable", () => VariableReference>
    export type VariableCreation = common.NamedIntrafile<"VariableCreation", common.RequiresOne<CustomTypeEntity> & common.RequiresOne<Assignable>>
    export type Statement = common.BaseStatement<() => ReturnStatement | Append | VariableCreation>

    export type FunctionBody = common.BaseFunctionBody<Statement>
    export type UnaryParameterType = common.PolymorphicEntity<"UnaryParameterType", () => CustomTypeEntity >
    export type NoParameter = common.IntrafileEntity<"NoParameter", {}>
    export type UnaryParameter = common.BaseUnaryParameter<UnaryParameterType>
    export type Parameter = common.PolymorphicEntity<"Parameter", () => UnaryParameter| NoParameter> 
    export type ReturnTypeSpec = common.BaseReturnTypeSpec<() => common.VoidReturn | CustomTypeEntity >
    export type Function = common.BaseFunction<FunctionBody, ReturnTypeSpec, Parameter>
    export type Struct = common.BaseStruct<Field>
    
    export type StoreDefinition = common.NamedIntrafile<"StoreDefinition", common.RequiresOne<CustomTypeEntity>>

    const symbolRegex: RegExp = new RegExp(`^(${Object.values(Symbol).join("|")})$`)

    type MatchResult = {hit: true, match: RegExpExecArray, loc: common.EntityLocation} | {hit: false}
    class FileCursor {
        private absOffset=0
        private line=0
        private column=0
        private readonly contents: string
        readonly filelocation: FileLocation
    
        constructor(contents: string, location: FileLocation) {
            this.contents = contents
            this.filelocation = location
        }
        
        isDone(): boolean {
            return this.absOffset >= this.contents.length
        }
    
        tryMatch(regex: RegExp): MatchResult {
            if (!regex.source.startsWith("^")) {
                throw new Error(`${regex.source} must match only the start of the string`)
            }
            const match = regex.exec(this.contents.slice(this.absOffset))
            if (match !== null) {
                const startLineNumber = this.line
                const startColNumber = this.column

                if(match.groups && match.groups.name && symbolRegex.test(match.groups.name)) {
                    throw new Error(`Entities may not be named after reserved symbols: Line: ${startLineNumber}\n\n${match[0]}`)

                }
                const matchStr = match[0]
                for(let i = 0; i < matchStr.length; ++i) {
                    if (matchStr[i] === "\n") {
                        this.line += 1
                        this.column = 0
                    } else {
                        this.column += 1
                    }
                }
                this.absOffset += matchStr.length
                return {
                    hit: true,
                    match,
                    loc: {
                        startColNumber,
                        startLineNumber,
                        endColNumber: this.column,
                        endLineNumber: this.line
                    }
                }
    
            } else {
                return {hit: false}
            }
        }

        getPositionHint(): string {
            return this.contents.slice(this.absOffset, this.absOffset + 20)
        }
    }

        
    export function extractAllFileEntities(contents: string, location: FileLocation): File {
        const cursor = new FileCursor(contents, location)
        const children = extractChildren<"File">(cursor, completeParserV2, {Enum: true, Struct: true, Function: true, StoreDefinition: true})
        if (cursor.tryMatch(/^\s*/).hit && cursor.isDone()) {
            return {
                kind: "File",
                loc: cursor.filelocation,
                children
            }
        }
        throw Error(`Failed to parse file entirely: ${JSON.stringify(location)}\n${cursor.getPositionHint()}`) 
    }

    type EntityOf<K extends WithChildren["kind"]> = Extract<WithChildren, {kind: K}>
    
    function extractChildren<K extends WithChildren["kind"]>(cursor: FileCursor, parserSet: CompleteParserV2, accepts: ChildrenDescription<EntityOf<K>>): EntityOf<K>["children"] {
        let tryExtractChild = true 
        const children: any = {}
        for (const k in accepts) {
            children[k] = []
        }
    
        while (tryExtractChild) {
            tryExtractChild = false
            for (const key in accepts) {
                
                const child = tryExtractEntity(cursor, 
                    key as keyof CompleteParserV2, 
                    parserSet)
                if (child !== undefined) {
                    tryExtractChild = true
                    
                    children[key].push(child)
                    break
                }
            }
        }
        return children
    }


    function extractToCompositeEntity<P extends AggregationParserV2<any>>(cursor: FileCursor, parser: P, parserSet: CompleteParserV2): Exclude<AnyEntity, File> | undefined {
        const m = cursor.tryMatch(parser.startRegex)
        if (!m.hit) {
            return undefined
        }
        
        const children = extractChildren(cursor, parserSet, parser.hasMany)
        const end = cursor.tryMatch(parser.endRegex)
        if (end.hit) {
            return parser.assemble(m.match, end.match, m.loc, children)
        }

        throw new Error(`Unable to parse end for entity: ${parser.endRegex.source}\n${cursor.getPositionHint()}`)
    }

    type AnyEntity = 
        File | 
        Struct | 
        Field | 
        common.Enum | 
        common.EnumMember | 
        FieldType | 
        CustomTypeEntity | 
        Function |
        FunctionBody |
        ReturnTypeSpec |
        Parameter | 
        common.VoidReturn |
        ReturnStatement | 
        Statement |
        UnaryParameterType |
        NoParameter |
        UnaryParameter | 
        StoreDefinition |
        Append |
        VariableReference |
        Returnable |
        Nothing |
        Assignable |
        VariableCreation

    type WithChildren = Extract<AnyEntity, {children: any}>
    type WithDependentClause= Extract<AnyEntity, {part: any}>


    function tryExtractEntity<K extends keyof ParserMap>(cursor: FileCursor, kind: K, parserSet: ParserMap): Exclude<AnyEntity, File> | undefined {
        const parser = parserSet[kind] as AggregationParserV2<any> | LeafParserV2<any> | ConglomerateParserV2<any> | PolymorphParser<any>
        switch(parser.kind) {
            case "aggregate":
                
                return extractToCompositeEntity(cursor, 
                    parser, 
                    parserSet)

            case "leaf":
                const match = cursor.tryMatch(parser.regex)
                if (match.hit) {
                    return parser.assemble(match.match, match.loc)
                }
                return undefined

            case "conglomerate":
                
                const start = cursor.tryMatch(parser.startRegex)
                if (!start.hit) {
                    return undefined
                }
                const part: any = {}
                const orderableMap: any = {}
                Object.entries(parser.requiresOne).forEach((v) => orderableMap[v[0]] = v[1].order)
                new Ordering(orderableMap).order.forEach(req => {
                    const childdef = parser.requiresOne[req]
                    if (childdef.beforeRegex !== undefined) {
                        if (!cursor.tryMatch(childdef.beforeRegex).hit) {
                            throw new Error(`Unable to parse prefix of ${req} for ${kind}`)
                        }
                    }

                    const depMatch = tryExtractEntity(cursor, req, parserSet)
                    if (depMatch === undefined) {
                        throw new Error(`Unable to parse required ${req} entity at ${JSON.stringify(start.loc)}\n\n ${cursor.getPositionHint()}`)
                    }
                    if (childdef.afterRegex !== undefined) {
                        if (!cursor.tryMatch(childdef.afterRegex).hit) {
                            throw new Error(`Unable to parse suffix of ${req} for ${kind}`)
                        }
                    }
                    part[req] = depMatch
                })

                const end = cursor.tryMatch(parser.endRegex)
                if (!end.hit) {
                    throw new Error(`Unable to find end of entity for ${kind} at ${cursor.getPositionHint()}`)
                }
                return parser.assemble(start.match, end.match, start.loc, part)

            case "polymorph":
                const order = new Ordering(parser.priority).order 
                for (let i = 0; i < order.length; i++) {
                    const elt = order[i];
                    const ent = tryExtractEntity(cursor, 
                        elt, 
                        parserSet)
                    if (ent !== undefined) {
                        return {kind: parser.groupKind, differentiate:() => ent as any}
                    }
                }
                return undefined

                
            default: assertNever(parser)

        }
        
    }

    type ChildrenDescription<K extends WithChildren> = Record<keyof K["children"], true>

    class Ordering<K extends keyof CompleteParserV2> {
        readonly order: K[]

        constructor(priorityMap: Record<K, number>) {
            const o = Object.entries(priorityMap) as [K, number][]
            this.order = o.sort((a, b) => a[1] - b[1]).map(a => a[0])
        }
    }

    type AggregationParserV2<K extends WithChildren> = Readonly<{
        kind: "aggregate"
        startRegex: RegExp
        assemble(start: RegExpExecArray, end: RegExpExecArray, loc: common.EntityLocation, children: K["children"]): K | undefined
        endRegex: RegExp
        hasMany: ChildrenDescription<K>,
    }>

    type LeafParserV2<K extends AnyEntity> = Readonly<{
        kind: "leaf"
        regex: RegExp
        assemble(c: RegExpExecArray, loc: common.EntityLocation): K | undefined
    }>

    type ConglomerateChildParseDefinition = Readonly<{
        beforeRegex?: RegExp
        afterRegex?: RegExp
        order: number
    }>
    type ConglomerateParserV2<K extends WithDependentClause> = Readonly<{
        kind: "conglomerate"
        startRegex: RegExp
        assemble(start: RegExpExecArray, end: RegExpExecArray, loc: common.EntityLocation, part: K["part"]): K | undefined
        endRegex: RegExp
        requiresOne: Record<Extract<keyof CompleteParserV2, keyof K["part"]>, ConglomerateChildParseDefinition>
    }>

    type PolymorphicEntity = Extract<AnyEntity, {differentiate(): any}>
    type PolymorphParser<K extends PolymorphicEntity> = {
        kind: "polymorph"
        // Ordering types perform the sort at startup. 
        // We use an object rather than all possible orderings of the kinds due to limitations of typescript.
        // The best we could do in typescript is an array of our union of kinds.
        // This is undesirable because you can compile a polymorphic type that hasn't prioritized all of its implementations.
        // Typescript does not have a way to go from a union to all possible ordering of union members, which is what we would want.
        // Further reading: https://github.com/Microsoft/TypeScript/issues/13298 
        // More reading: https://github.com/microsoft/TypeScript/issues/26223#issuecomment-513187373
        priority: Record<Extract<keyof CompleteParserV2, ReturnType<K["differentiate"]>["kind"]>, number>
        groupKind: K["kind"]
    }

    type ToFullEntity<K extends common.EntityKinds> = Extract<AnyEntity, {kind: K}>
    type SelectParserType<E extends AnyEntity> = E extends WithChildren ? AggregationParserV2<E> : (
        E extends WithDependentClause ? ConglomerateParserV2<E> : 
            E extends PolymorphicEntity ? PolymorphParser<E> :
                E extends Exclude<AnyEntity, WithDependentClause | WithChildren> ? LeafParserV2<E> : never
    )

    type GetAllDependencies<E extends keyof ParserMap> = E extends WithChildren["kind"] ? keyof Extract<WithChildren, {kind: E}>["children"] :
        E extends WithDependentClause["kind"] ? keyof Extract<WithDependentClause, {kind: E}>["part"] : 
            E extends PolymorphicEntity["kind"] ? ReturnType<Extract<Extract<PolymorphicEntity, {differentiate: any}>, {kind: E}>["differentiate"]>["kind"] : never
    
    type ParserMap = {
        [P in Exclude<AnyEntity, File>["kind"]]:  SelectParserType<ToFullEntity<P>>
    }

    type CompleteParserV2 = ParserMap & {
        [P in keyof ParserMap]: {
            [Q in Exclude<GetAllDependencies<P>, keyof ParserMap>]: Q extends never ? {} : "This entity needs to be added to the AnyEntity union"
        }
    }

    const completeParserV2: CompleteParserV2 = {
        Enum: {
            kind: "aggregate",
            startRegex: /^\s*enum +(?<name>[a-zA-Z_]\w*) *{/,
            assemble(start, end, loc, children): common.Enum | undefined {
                return {
                    kind: "Enum",
                    name: start.groups.name,
                    loc,
                    children
                }
            },
            endRegex:/^\s*}/,
            hasMany: {EnumMember: true}
        },
        
        EnumMember: {
            kind: "leaf",
            regex: /^\s*(?<name>[a-zA-Z_]\w*)(,|\s)/,
            assemble(c, loc): common.EnumMember | undefined {

                return{
                    kind: "EnumMember",
                    name: c.groups.name,
                    loc
                }   
            }
        },


        FieldType: {
            kind: "polymorph",
            priority: {
                CustomType: 2
            },
            groupKind: "FieldType"
        },

        Field: {
            kind: "conglomerate",
            startRegex: /^\s*(?<name>[_A-Za-z]+[\w]*): */,
            endRegex: /^\s*(,|\n|(?= *}))/,
            assemble(start, end, loc, part): Field | undefined {
                return {
                    kind: "Field",
                    name: start.groups.name,
                    loc,
                    part
                }
            },
            requiresOne: {
                FieldType: {order: 1}
            },
        },

        Struct: {
            kind: "aggregate",
            startRegex: /^\s*struct +(?<name>[a-zA-Z_]\w*) *{/,
            assemble(start, end, loc, children): Struct | undefined {
                return {
                    kind: "Struct",
                    name: start.groups.name,
                    loc,
                    children
                }
            },
            endRegex:/^\s*}/,
            hasMany: {Field: true}
        },
        CustomType: {
            kind: "leaf",
            regex: /^ *((?<modifier>Array|Optional)<)? *(?<type>[_A-Za-z]+[\w]*) *(?<closer>>)?/,
            assemble(match, loc): CustomTypeEntity | undefined {
                let modification: common.TypeModification = "none"
                if (match.groups.modifier && !match.groups.closer) {
                    return undefined
                }

                if (match.groups.modifier === "Array") {
                    modification = "array"
                } else if (match.groups.modifier === "Optional") {
                    modification = "optional"
                }
                return {
                    kind: "CustomType",
                    loc,
                    type: match.groups.type,
                    modification
                }
            }
        },
        Function: {
            kind: "conglomerate",
            startRegex: /^\s*function +(?<name>[a-zA-Z_]\w*)/,
            endRegex: /^/,
            assemble(start, end, loc, part): Function | undefined {
                return {
                    kind: "Function",
                    loc,
                    name: start.groups.name,
                    part
                }
            },
            requiresOne: {
                FunctionBody: {order: 3}, 
                Parameter: {order: 1}, 
                ReturnTypeSpec: {
                    beforeRegex: /^ *:?/,
                    order: 2
                }
            }
        },
        ReturnTypeSpec: {
            kind: "polymorph",
            groupKind: "ReturnTypeSpec",
            priority: {CustomType: 2, VoidReturnType: 3}
        },
        Parameter: {
            kind: "polymorph",
            groupKind: "Parameter",
            priority: { UnaryParameter: 2, NoParameter: 1}
        },

        NoParameter: {
            kind: "leaf",
            regex: /^\(\)/,
            assemble(c, loc) {
                return {
                    kind: "NoParameter",
                    loc
                }
            }
        },

        UnaryParameter: {
            kind: "conglomerate",
            startRegex: /^\(\s*(?<name>[a-zA-Z_]\w*): */,
            endRegex: /^\s*\) */,
            requiresOne: {UnaryParameterType: {order: 1}},
            assemble(start, end, loc, part) {
                return {
                    kind: "UnaryParameter",
                    name: start.groups.name,
                    loc, 
                    part,
                }
            }

        },

        FunctionBody: {
            kind: "aggregate",
            startRegex: /^\s*{/,
            endRegex: /^\s*}/,
            assemble(start, end, loc, children): FunctionBody | undefined {
                return {
                    kind: "FunctionBody",
                    loc,
                    children
                }
            },
            hasMany: {Statement: true}
        },
        UnaryParameterType: {
            kind: "polymorph",
            priority: {CustomType: 2},
            groupKind: "UnaryParameterType"
        },
        VoidReturnType: {
            kind: "leaf",
            regex: /^\s*(?=\{)/,
            assemble(c, loc) {
                return {
                    kind: "VoidReturnType",
                }
            }
        },
        Statement: {
            kind: "polymorph",
            groupKind: "Statement",
            priority: {ReturnStatement: 1, Append: 2, VariableCreation: 3}
        },
        ReturnStatement: {
            kind: "conglomerate",
            startRegex: /^\s*return +/,
            endRegex: /^/,
            requiresOne: {
                Returnable: {order: 1}
            },
            assemble(start, end, loc, part) {
                return {
                    kind: "ReturnStatement",
                    loc,
                    part
                }
            }
        },
        StoreDefinition: {
            kind: "conglomerate",
            startRegex: /^\s*(?<name>[a-zA-Z]+):\s*/,
            endRegex: /^\s*=\s*\[\]/,
            requiresOne: {
                CustomType: {order: 1}
            },
            assemble(start, end, loc, part) {
                return {
                    kind: "StoreDefinition",
                    loc: loc,
                    name: start.groups.name,
                    part
                }
            }
        },
        Append: {
            kind: "leaf",
            regex: /^\s*(?<storeName>[a-zA-Z]+)\.append\(\s*(?<variableName>[a-zA-Z_]\w*)\s*\)\s*/,
            assemble(c, loc) {
                return {
                    kind: "Append",
                    loc,
                    variableName: c.groups.variableName,
                    storeName: c.groups.storeName
                }
            }
        },
        VariableReference: {
            kind: "leaf",
            regex: /^\s*(?<name>[a-zA-Z_]\w*)/,
            assemble(c, loc) {
                return {
                    kind: "VariableReference",
                    loc,
                    val: c.groups.name
                }
            }
        },
        Returnable: {
            kind: "polymorph",
            priority: {
                Assignable: 1,
                Nothing: 2
            },
            groupKind: "Returnable"
        },
        Nothing: {
            kind: "leaf",
            regex: /^/,
            assemble() {
                return {
                    kind: "Nothing"
                }
            }
        },
        VariableCreation: {
            kind: "conglomerate",
            startRegex: /^\s*(?<name>[a-zA-Z_]\w*) *:/,
            endRegex: /^/,
            assemble(start, end, loc, part) {
                return {
                    kind: "VariableCreation",
                    name: start.groups.name,
                    part,
                    loc
                }
            },
            requiresOne: {
                CustomType: {
                    afterRegex: /^ *=/,
                    order: 1
                },
                Assignable: {
                    order: 2
                }
            }
        },
        Assignable: {
            kind: "polymorph",
            priority: {
                VariableReference: 1
            },
            groupKind: "Assignable"
        }
    }
}