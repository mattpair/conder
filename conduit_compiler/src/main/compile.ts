
import { Parse} from "./parse";
import { FunctionResolved, TypeResolved, Message, Field } from "./entity/resolved";
import { resolveDeps } from "./resolution/resolveTypes";
import { FileLocation } from "./util/filesystem";
import { Enum, EnumMember } from "./entity/basic";
import { resolveFunctions } from "./resolution/resolveFunction";

export type PartiallyCompiled = Record<string, {proto: string, data: FunctionResolved.File}>
export function compileFiles(files: Record<string, () => string>): PartiallyCompiled {
    const conduits: Parse.File[] = []
    for (const file in files) {
        if (file.endsWith(".cdt")) {
            conduits.push(Parse.extractAllFileEntities(files[file](), new FileLocation(file)))
        }
    }

    const r = resolveDeps(conduits)

    return resolveFunctionsAndProto(r)
} 

function resolveFunctionsAndProto(files: TypeResolved.File[]): PartiallyCompiled {
    const results: PartiallyCompiled = {}
    files.forEach(file => {

        const enums: Enum[] = []
        const messages: Message[] = []
        file.inFileScope.forEach(v => {
            switch(v.kind) {
                case "Enum":
                    enums.push(v)
                    break;
                case "Message":
                    messages.push(v)
                    break;
            }
        })
        let proto = ''
        if (enums.length + messages.length > 0) {
            proto = `
syntax="proto2";
${file.children.Import.map(d => `import "${d.dep.replace(".cdt", ".proto")}";`).join("\n")}
            
${enums.map(printEnum).join("\n\n")}
${messages.map(printMessage).join("\n\n")}
`
        }

        results[`${file.loc.fullname.replace(".cdt", ".proto")}`] = {proto, data: resolveFunctions(file)}
        
    })

    return results       
}

function printEnum(e: Enum): string {
    const mems = printMembers(e.children.EnumMember)
    return `
enum ${e.name} {
${mems}
}
    `
}

function printMembers(m: EnumMember[]): string {
    return m.map((e, index) => `\t${e.name} = ${index + 1};`).join("\n")
}

function printMessage(m: Message): string {
    const fields = printFields(m.children.Field)

    return `
message ${m.name} {
${fields}
}`
}

function printFields(fields: Field[]): string {
    return fields
    .map((f, index) => {
        const p = f.part.FieldType.differentiate()
        return `\t${f.isRequired ? 'required' : 'optional'} ${p.kind === "Primitive" ? p.val : p.name} ${f.name} = ${index + 1};`
    })
    .join("\n")
}