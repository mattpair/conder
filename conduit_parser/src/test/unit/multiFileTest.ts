import {compileFiles} from '../../main/compile'

function makeLazy(files: Record<string, string>): Record<string, () => string> {
    const lazyFiles: Record<string, () => string> = {}
        for (const file in files) {
            lazyFiles[file] = () => files[file]
        }
    return lazyFiles
}

function protoCompileTest(description: string, files: Record<string, string>) {
    test(description, () => {

        expect(compileFiles(makeLazy(files), {dependents: {}, project: "test", install: []})).toMatchSnapshot()
    })
} 

function testFailsWhen(description: string, files: Record<string, string>) {
    test(description, () => {
        expect(() => compileFiles(makeLazy(files),{dependents: {}, project: "test", install: []})).toThrowErrorMatchingSnapshot()
    })
}


protoCompileTest("simple multi file", {
    "conduit_a.cdt": `
    struct m1 {
        d: double
    }
    `,

    "conduit_b.cdt": `

    struct m2 {
        d: double
    }
    struct m3 {
        m: m2
    }
    `
})

protoCompileTest("dependency multi file", {
    "conduit_a.cdt": `
    struct m1 {
        d: double
    }
    `,

    "conduit_b.cdt": `

    struct m2 {
        m:m1
    }
    `
})

testFailsWhen("Self referencing type", {
    "conduit_a.cdt": `
    struct m1 {
        d: m1
    }
    `,

})

testFailsWhen("function with unknown types", {
    "conduit_a.cdt": `
    function letsGetFuncy(a: SomeType, b: Foreign.Type) {

    }

    `
})

testFailsWhen("function with invalid return type", {
    "conduit_a.cdt": `

    function funk() SomeType {
    }
    `
})

protoCompileTest("simple echo function", {
    "conduit_a.cdt": `

    struct Shout {
        m: string
    }

    function echo(s: Shout) Shout {
        return s
    }
    `
})
