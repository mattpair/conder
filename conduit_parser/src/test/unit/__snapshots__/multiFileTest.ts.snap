// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Self referencing type 1`] = `"Currently do not support self-referencing types: m1"`;

exports[`dependency multi file 1`] = `
Object {
  "inScope": EntityMap {
    "map": Map {
      "m1" => Object {
        "children": Object {
          "Field": Array [
            Object {
              "kind": "Field",
              "loc": Object {
                "endColNumber": 11,
                "endLineNumber": 2,
                "startColNumber": 15,
                "startLineNumber": 1,
              },
              "name": "d",
              "part": Object {
                "FieldType": Object {
                  "differentiate": [Function],
                  "kind": "FieldType",
                },
              },
            },
          ],
        },
        "file": FileLocation {
          "dir": "",
          "fullname": "conduit_a.cdt",
          "name": "conduit_a.cdt",
        },
        "kind": "Struct",
        "loc": Object {
          "endColNumber": 15,
          "endLineNumber": 1,
          "startColNumber": 0,
          "startLineNumber": 0,
        },
        "name": "m1",
      },
      "m2" => Object {
        "children": Object {
          "Field": Array [
            Object {
              "kind": "Field",
              "loc": Object {
                "endColNumber": 10,
                "endLineNumber": 3,
                "startColNumber": 15,
                "startLineNumber": 2,
              },
              "name": "m",
              "part": Object {
                "FieldType": Object {
                  "differentiate": [Function],
                  "kind": "FieldType",
                },
              },
            },
          ],
        },
        "file": FileLocation {
          "dir": "",
          "fullname": "conduit_b.cdt",
          "name": "conduit_b.cdt",
        },
        "kind": "Struct",
        "loc": Object {
          "endColNumber": 15,
          "endLineNumber": 2,
          "startColNumber": 0,
          "startLineNumber": 0,
        },
        "name": "m2",
      },
    },
    "size": 2,
  },
}
`;

exports[`function with invalid return type 1`] = `"Could not find a Struct named SomeType"`;

exports[`function with unknown types 1`] = `"Unable to find end of entity for UnaryParameter at , b: Foreign.Type) {"`;

exports[`simple echo function 1`] = `
Object {
  "inScope": EntityMap {
    "map": Map {
      "Shout" => Object {
        "children": Object {
          "Field": Array [
            Object {
              "kind": "Field",
              "loc": Object {
                "endColNumber": 11,
                "endLineNumber": 3,
                "startColNumber": 18,
                "startLineNumber": 2,
              },
              "name": "m",
              "part": Object {
                "FieldType": Object {
                  "differentiate": [Function],
                  "kind": "FieldType",
                },
              },
            },
          ],
        },
        "file": FileLocation {
          "dir": "",
          "fullname": "conduit_a.cdt",
          "name": "conduit_a.cdt",
        },
        "kind": "Struct",
        "loc": Object {
          "endColNumber": 18,
          "endLineNumber": 2,
          "startColNumber": 0,
          "startLineNumber": 0,
        },
        "name": "Shout",
      },
      "echo" => Object {
        "body": Object {
          "kind": "FunctionBody",
          "loc": Object {
            "endColNumber": 35,
            "endLineNumber": 6,
            "startColNumber": 34,
            "startLineNumber": 6,
          },
          "statements": Array [
            Object {
              "differentiate": [Function],
              "kind": "Statement",
            },
          ],
        },
        "kind": "Function",
        "loc": Object {
          "endColNumber": 17,
          "endLineNumber": 6,
          "startColNumber": 5,
          "startLineNumber": 4,
        },
        "method": "POST",
        "name": "echo",
        "parameter": Object {
          "differentiate": [Function],
          "kind": "Parameter",
        },
        "returnType": Object {
          "kind": "CustomType",
          "modification": "none",
          "type": "Shout",
        },
      },
    },
    "size": 2,
  },
}
`;

exports[`simple multi file 1`] = `
Object {
  "inScope": EntityMap {
    "map": Map {
      "m1" => Object {
        "children": Object {
          "Field": Array [
            Object {
              "kind": "Field",
              "loc": Object {
                "endColNumber": 11,
                "endLineNumber": 2,
                "startColNumber": 15,
                "startLineNumber": 1,
              },
              "name": "d",
              "part": Object {
                "FieldType": Object {
                  "differentiate": [Function],
                  "kind": "FieldType",
                },
              },
            },
          ],
        },
        "file": FileLocation {
          "dir": "",
          "fullname": "conduit_a.cdt",
          "name": "conduit_a.cdt",
        },
        "kind": "Struct",
        "loc": Object {
          "endColNumber": 15,
          "endLineNumber": 1,
          "startColNumber": 0,
          "startLineNumber": 0,
        },
        "name": "m1",
      },
      "m2" => Object {
        "children": Object {
          "Field": Array [
            Object {
              "kind": "Field",
              "loc": Object {
                "endColNumber": 11,
                "endLineNumber": 3,
                "startColNumber": 15,
                "startLineNumber": 2,
              },
              "name": "d",
              "part": Object {
                "FieldType": Object {
                  "differentiate": [Function],
                  "kind": "FieldType",
                },
              },
            },
          ],
        },
        "file": FileLocation {
          "dir": "",
          "fullname": "conduit_b.cdt",
          "name": "conduit_b.cdt",
        },
        "kind": "Struct",
        "loc": Object {
          "endColNumber": 15,
          "endLineNumber": 2,
          "startColNumber": 0,
          "startLineNumber": 0,
        },
        "name": "m2",
      },
      "m3" => Object {
        "children": Object {
          "Field": Array [
            Object {
              "kind": "Field",
              "loc": Object {
                "endColNumber": 11,
                "endLineNumber": 6,
                "startColNumber": 15,
                "startLineNumber": 5,
              },
              "name": "m",
              "part": Object {
                "FieldType": Object {
                  "differentiate": [Function],
                  "kind": "FieldType",
                },
              },
            },
          ],
        },
        "file": FileLocation {
          "dir": "",
          "fullname": "conduit_b.cdt",
          "name": "conduit_b.cdt",
        },
        "kind": "Struct",
        "loc": Object {
          "endColNumber": 15,
          "endLineNumber": 5,
          "startColNumber": 5,
          "startLineNumber": 4,
        },
        "name": "m3",
      },
    },
    "size": 3,
  },
}
`;
